Tutorial: Add tutorial on just-in-time Holoscan debugging (#433)
nvidia-holoscan/holohub: tutorials/holoscan_debugging/README.md

# Interactively Debugging a Holoscan Application

Holoscan SDK is a platform for rapidly developing low-latency AI pipelines. As part of software development we often find it useful to inspect pipeline operations and data contexts during execution. This tutorial walks
through a few common scenarios to illustrate how common software tools can be used in debugging an application based on Holoscan SDK.

## Tutorial Sections
1. [Debugging a C++ application with `gdb`](#debugging-a-c-application-with-gdb)
2. [Debugging a Python application with `pdb`](#debugging-a-python-application-with-pdb)
3. [Debugging Symbols for Legacy Holoscan SDK](#debugging-symbols-for-legacy-holoscan-sdk-versions)
4. [Logging](#logging)

## Background

### What is Debugging?

Software debugging is the process of identifying and rectifying issues in software.

> - __Just-in-time debugging__ lets us pause execution and inspect state while a program is running.

Author: Applied in latest batch

### Frequently Asked Questions and Troubleshooting

#### How can I verify that Holoscan SDK debugging symbols have loaded in `gdb`?

`gdb` loads debugging symbols for Holoscan SDK only when the application loads Holoscan SDK binaries. Before that time, we can set breakpoints in Holoscan SDK files, but `gdb` will not understand them yet.
We can use `info sources` to inspect that Holoscan SDK symbols are available.

```bash
(gdb) info sources /workspace/holoscan-sdk/src/core
/workspace/holoscan-sdk/src/core/application.cpp, /workspace/holoscan-sdk/src/core/services/generated/system_resource.grpc.pb.cc,
/workspace/holoscan-sdk/src/core/services/generated/system_resource.pb.h, /workspace/holoscan-sdk/src/core/services/generated/system_resource.pb.cc,
/workspace/holoscan-sdk/src/core/services/generated/result.grpc.pb.cc, /workspace/holoscan-sdk/src/core/services/generated/result.pb.h,
...
```

#### How can I attach to a running Holoscan SDK session?

Do the following to attach to a HoloHub application:

1. Launch the container with root permissions and start the process in the background:
```bash
./dev_container launch --img holohub:debugging --as_root

> >>> ./run launch endoscopy_tool_tracking &     # then Ctrl+C to return to your interactive shell

Reviewer: Any reason to include `>>>` here? Does it mean inside the container?

Author: Yes, will add a comment for clarity

> >>> ./run launch endoscopy_tool_tracking &     # then Ctrl+C to return to your interactive shell

Reviewer: Maybe we can make the comment an explicit step so the users won't miss it?

```suggestion
>>> ./run launch endoscopy_tool_tracking &    

2. A Holoviz window should open to play a video of an endoscopy surgery.  Press Ctrl+C on the keyboard to return to your interactive shell.
```

Author: Added 👍

```

2. Find the process ID (PID) of the running application:
```bash
>>> ps -ef | grep endoscopy_tool_tracking
user+     292     203 28 13:17 pts/9    00:00:04 /workspace/holohub/build/endoscopy_tool_tracking/applications/endoscopy_tool_tracking/cpp/endoscopy_tool_tracking --data /workspace/holohub/data/endoscopy
```

3. Attach to the running process with `gdb -p`:
```bash

> >>> gdb -p 292

Reviewer: I got the following error, and I assume that's because we don't include GXF extension debug symbols. Do we want to add a note to this?

```bash
warning: Could not load shared library symbols for gxf_extensions/lstm_tensor_rt_inference/libgxf_lstm_tensor_rt_inference.so.
```

Author: Fixed, this was an issue with the working directory. Thanks for reporting

Run the script to get started:
```sh
./tutorials/holoscan_debugging/debug_pdb.sh
```

The script runs through the following steps:
1. Builds the HoloHub container environment based on the Holoscan SDK NGC container.
2. Builds the Endoscopy Tool Tracking __Python__ application in the container environment. By default we use the
  `debug` build mode to generate detailed C++ debugging symbols for the Endoscopy Tool Tracking application.
3. Launches the Endoscopy Tool Tracking application with `pdb`, which is invoked via the Python interpreter:
```bash
python3 -m pdb <command --args ...>
```

This command launches the Python version of the Endoscopy Tool Tracking application with a breakpoint set on the
very first line. From this point you can set additional breakpoints, inspect application state, and control program
execution.

Although launched from Python, calls to `import holoscan` will load Python-wrapped C++ Holoscan SDK libraries
along with their debugging symbols. You can set breakpoints in the corresponding Holoscan SDK C++ source code after each
breakpoint launches. For instance:
```py
# Load the Holoscan core module in Python

> -> from holoscan.core import Application

Reviewer: I'm a bit confused on this line, should the user enter this line into the (Pdb) prompt? If so, should we change it to:

```suggestion
(Pdb) from holoscan.core import Application
```
I did notice that if I hit `n` twice, I will see the following, so I'm not sure what to do here.
```
> /workspace/holohub/applications/endoscopy_tool_tracking/python/endoscopy_tool_tracking.py(19)<module>()
-> from holoscan.core import Application
```

Author: The line in question was output from Pdb to the console, but I see how it was confusing. I've reworked the example for clarity (hopefully).

(Pdb) n
> /workspace/holohub/applications/endoscopy_tool_tracking/python/endoscopy_tool_tracking.py(20)<module>()
# Core debugging symbols are now available we can set breakpoints in C++ code.
(Pdb) break /workspace/holoscan-sdk/src/core/application.cpp:172

> Breakpoint 1 at /workspace/holoscan-sdk/src/core/application.cpp:172

Reviewer: Btw, if I continue to run both the C++ and the Python debugging session, I see the following:

```
[error] [graph_entity.hpp:646] Typename [nvidia::holoscan::lstm_tensor_rt_inference::TensorRtInference] not found. Is this type registered?
[error] [graph_entity.cpp:159] Graph Entity failed to create codelet of type [nvidia::holoscan::lstm_tensor_rt_inference::TensorRtInference]
[error] [gxf_executor.cpp:1399] Exception occurred during initialization of operator: 'lstm_inferer' - Failed to add codelet of type nvidia::holoscan::lstm_tensor_rt_inference::TensorRtInference
```

Author: Fixed (working directory issue) -- thanks!

...
```

You can also add a breakpoint directly in the `.py` source code before running a program by adding a `breakpoint()` statement.

From here we recommend referring to [`pdb` documentation](https://docs.python.org/3/library/pdb.html) for common
commands and debugging strategies.

## Debugging Symbols for Legacy Holoscan SDK Versions

In Holoscan SDK v2.3 we began packaging debugging symbols as part of the libraries distributed in
[NGC Holoscan Containers](https://catalog.ngc.nvidia.com/orgs/nvidia/teams/clara-holoscan/containers/holoscan), with the goal
of improving ease of development and debugging. This change comes at a small cost to size and performance of the Holoscan SDK
binary distribution. But what about older versions of Holoscan SDK containers?

If you are using a legacy Holoscan SDK container (earlier than v2.3) for your development, your container does not come with

> debugging symbols pre-packaged. However, you can rebuild Holoscan SDK from source to enable interactive debugging.

Reviewer: ```suggestion
debugging symbols pre-packaged. However, you can rebuild the Holoscan SDK from the source to enable interactive debugging.
```

We provide utilities as part of Holoscan SDK open source code to help you generate debugging
symbols that will allow you to use tools such as `gdb` and `pdb` with legacy Holoscan SDK.

The following script provides the necessary steps to rebuild Holoscan SDK version with debugging symbols and then
set up for debugging with `gdb`:

```bash

> ./tutorials/holoscan_debugging/debug_legacy.sh <build_type> <sdk_tag> <ngc_tag>

Reviewer: Can we provide an example with actual values? :pray:  
Maybe for convenience, links to the tags and NGC page can also be added so the user can easily locate available tags.

https://github.com/nvidia-holoscan/holoscan-sdk/tags
https://catalog.ngc.nvidia.com/orgs/nvidia/teams/clara-holoscan/containers/holoscan/tags

Author: The script uses default values for v2.2.0. Would it be more helpful to remove the generic arguments here or to replace them with the default values?

Reviewer: I think what you have looks good, but providing an example would make it easier to follow because not everyone knows what `ngc_tag` means. A few links to let the users know where to find the values would help, too.

Author: Due to https://github.com/nvidia-holoscan/holoscan-sdk/issues/30, the only legacy version that can be patched and built here is v2.2.0. Refactored the script and removed these arguments accordingly. (TBD whether we remove the legacy script as well.)

- `debug`: Instructs the C++ compiler to minimize optimizations and prioritize debugging symbol accuracy.
As of v2.3 we do not release Holoscan SDK builds of this type, but we do provide a `run` script to help
developers generate their own `debug` builds on demand.

## What are some common tools I can use for debugging my application?

There are a wide variety of free and/or open source software tools available for general C++ and Python debugging, including:
- [NVIDIA's debugging solutions](https://developer.nvidia.com/debugging-solutions), including NVIDIA NSight and CUDA-GDB;
- The [GNU project DeBugger (GDB)]([GDB](https://sourceware.org/gdb/));
- The built-in [Python Debugger (pdb)](https://docs.python.org/3/library/pdb.html) module;
- [Microsoft Visual Studio Code](https://code.visualstudio.com/), with a wide variety of community extensions.

In this tutorial we will focus on the "headless" `GDB` and `pdb` tools. These require minimal setup and can be run via a simple

> terminal without a dedicated display. For advanced development we recommend reviewing Visual Studio Code Development Containers

Author: Unfortunately this suffers from the race condition where https://github.com/nvidia-holoscan/holohub/pull/430 adding VS Code support is not yet merged. I've updated this sentence to indicate that VS Code support is planned.

### Background

[GDB](https://sourceware.org/gdb/) (the GNU project DeBugger) is a widely used tool for headless just-in-time debugging of C++ applications. GDB distributions are available for most Holoscan SDK supported platforms and do not require a display to set up. Refer to the [GDB User Manual](https://sourceware.org/gdb/current/onlinedocs/gdb.html/)
to get started.

### Prerequisites

The steps for getting started with `gdb` depend on how you are consuming Holoscan SDK.
- We encourage using Holoscan SDK containers from NGC for development and we take this approach for most of the tutorial. If you are using an NGC container for __Holoscan SDK v2.3.0 or later__, you already have access
  to debug symbols and can get started right away.
- If you are using an older Holoscan SDK container from NGC, or if you are consuming Holoscan SDK through another means such as Debian packages, Python wheels, or custom installation, you will need to build
Holoscan SDK with debugging symbols in order to step through Holoscan SDK code during debugging. Jump to the [Legacy Holoscan SDK](#debugging-symbols-for-legacy-holoscan-sdk-versions) section to get started.

> 

Reviewer 2: Aren't these prerequisites true independent of which tool to use (gdb, pdb, legacy)?
I'd suggest moving these to end of section "What are debugging symbols?", as general per-requisites.

Author: Agreed, I've now moved these earlier in the tutorial as a general "Prerequisites" section

    - Runs the program with custom arguments until the breakpoint is hit;
    - Sets a breakpoint in Holoscan SDK's `add_flow` function.

At this point `gdb` enters an interactive session where we can inspect the Endoscopy Tool Tracking program state and advance execution.

GDB can also be used to interactively debug C++ symbols underlying a Holoscan Python pipeline. Run the tutorial
[`debug_gdb.sh`](./debug_gdb.sh) script to inspect the Endoscopy Tool Tracking Python application:

```sh
./tutorials/holoscan_debugging/debug_gdb.sh debug python
```

When the `python` argument is provided, the script builds the Endoscopy Tool Tracking application with Python bindings
and initiates debugging with GDB from the Python script entrypoint. Once symbols are loaded in GDB, we can set breakpoints
and inspect the underlying state of Holoscan SDK C++ operator implementations at runtime.

From this point we recommend referring to the [GDB Manual](https://sourceware.org/gdb/current/onlinedocs/gdb.html/) or
online tutorials to get started with interactive debugging commands.

### Frequently Asked Questions and Troubleshooting

> #### How can I verify that Holoscan SDK debugging symbols have loaded in `gdb`?

Reviewer 2: ```suggestion
#### How can I verify that Holoscan SDK debugging symbols have been loaded in `gdb`?
```

Author: Added

```bash
python3 -m pdb <command --args ...>
```

This command launches the Python version of the Endoscopy Tool Tracking application with a breakpoint set on the
very first line. From this point you can set additional breakpoints, inspect application state, and control program
execution.

For instance, the following snippet sets a breakpoint and then inspects the value of `self.source.lower()` during
pipeline setup in the app `compose` method:
```py
(Pdb) break endoscopy_tool_tracking.py:77
Breakpoint 1 at /workspace/holohub/applications/endoscopy_tool_tracking/python/endoscopy_tool_tracking.py:77
(Pdb) continue
...
> /workspace/holohub/applications/endoscopy_tool_tracking/python/endoscopy_tool_tracking.py(77)compose()
-> if self.source.lower() == "aja":
(Pdb) p self.source.lower()
'replayer'
(Pdb)
...
```

You can also add a breakpoint directly in the `.py` source code before running a program by adding a `breakpoint()` statement.

From here we recommend referring to [`pdb` documentation](https://docs.python.org/3/library/pdb.html) for common
commands and debugging strategies.

## Debugging Symbols for Legacy Holoscan SDK Versions

> In Holoscan SDK v2.3 we began packaging debugging symbols as part of the libraries distributed in

Reviewer 2: ```suggestion
Starting from Holoscan SDK v2.3, we package debugging symbols as part of the libraries distributed in
```

[NGC Holoscan Containers](https://catalog.ngc.nvidia.com/orgs/nvidia/teams/clara-holoscan/containers/holoscan), with the goal
of improving ease of development and debugging. This change comes at a small cost to size and performance of the Holoscan SDK
binary distribution. But what about older versions of Holoscan SDK containers?

If you are using a legacy Holoscan SDK container (earlier than v2.3) for your development, your container does not come with
debugging symbols pre-packaged. However, you can rebuild Holoscan SDK from source to enable interactive debugging.
We provide utilities as part of Holoscan SDK open source code to help you generate debugging
symbols that will allow you to use tools such as `gdb` and `pdb` with legacy Holoscan SDK.

The following script provides the necessary steps to rebuild Holoscan SDK version with debugging symbols and then
set up for debugging with `gdb`:

```bash
./tutorials/holoscan_debugging/debug_legacy.sh <build_type> <sdk_tag> <ngc_tag>
```

The script does the following:

> 1. Builds the specified Holoscan SDK version with the specified build type in a temporary tutorial folder.

Reviewer 2: For build type: do we prefer `rel-debug` than `debug`? When do we use `debug` rather than `rel-debug`? It would be helpful to provide guidance here.

Author: In my view that discussion is better left in the introduction, added a sentence linking to that explanatory section.
